### DISCUSSION TITLE
Building pnpm from source (without pnpm)

### DISCUSSION THREAD

**Discussion by taylorjdawson** [pnpm/pnpm#8602] [Q&A]
We are looking to build pnpm from source, however, it looks like you need pnpm to build pnpm. Is there a way for us to build pnpm without using pnpm? This is a boot strapping issue.

I couldn't locate in the repo the code for the `pnpm compile` command if someone could point me to where that is.

**Comment by zkochan** [comment]
No, I don't think you could. But you can do `npx pnpm install` and so on. So technically you wouldn't install pnpm just run `npx`.

**Comment by rastagraffix** [comment]
This all assumes there is already a pre-built package. For OpenBSD, there isn't. That's why we need to figure out how to build this monster from scratch.

**Reply by zkochan** [reply]
We could commit a compiled pnpm to the repository, I guess.

**Reply by rastagraffix** [reply]
Well, that would solve the immediate problem, I guess. There would need to be binaries for arm64 and amd64.

But that doesn’t solve the underlying problem: explaining how to build pnpm from scratch. If I knew how to do that, I could create an OpenBSD port for this that would merge into the ports tree, making it available to everyone running OpenBSD. Same for FreeBSD. And Solaris. Etc.

**Reply by zkochan** [reply]
you can download the pnpm artefacts from the [release page](https://github.com/pnpm/pnpm/releases/tag/v9.12.0) and use it to build pnpm.

**Reply by lrvick** [reply]
This would not be a solution at any of the orgs I support, or in the Linux distro I maintain, which all have supply chain integrity standards that forbid trusting third party compiled binaries. How would anyone know that binary you upload was not tampered with? 

There must be a way to bootstrap pnpm from source. How was the first version of pnpm built?

**Reply by rastagraffix** [reply]
I have already downloaded the source code. The issue is THERE ARE NO INSTRUCTIONS ON HOW TO BUILD IT.

I don't know how to make this any simpler or clearer.

**Reply by konsolebox** [reply]
@zkochan Why does building of pnpm have to be cryptic? Is there a step that shouldn't be publicized?

**Reply by zkochan** [reply]
> I have already downloaded the source code. The issue is THERE ARE NO INSTRUCTIONS ON HOW TO BUILD IT.
> I don't know how to make this any simpler or clearer.

You cannot build pnpm without using pnpm. How can I be any simpler or clearer?

**Reply by zkochan** [reply]
> This would not be a solution at any of the orgs I support, or in the Linux distro I maintain, which all have supply chain integrity standards that forbid trusting third party compiled binaries. How would anyone know that binary you upload was not tampered with?

The npm registry has added support for provenance and OIDC. We will configure that for pnpm

**Comment by orsolic** [comment]
One year later - is it possible to build pnpm from source (preferably on FreeBSD)?

**Reply by zkochan** [reply]
No, you cannot build pnpm without using pnpm.

**Reply by theoparis** [reply]
That is unfortunate, as I need to use a custom node.js build with patches and I need to bootstrap pnpm somehow. I guess I could download pre-built dist tarballs from npm directly and then add a shebang to pnpm.cjs.

**Comment by geaaru** [comment]
hi,
i just add my cent.
It seems clear that the current release and building depends on itself but like for Golang for example, but it sounds interesting understand how it's been build the first release and what are the packages patches to build it. I don't find any description in the website about this step or if it exists please share the link here.
This could be helpful to understand the story and help the community to eventually create a way to build from source this tool for example with a bootstrap tarball that could be created. A lot of distro source-based consider this process a lot important to ensure a good security level and a reproducible way to create a tree.
Do you think that could be possible using an old release to build from source the bootstrap tarball that is later used to build the last release?
If it's not possible, can you add a description in the website that describe the step used to arrive at the first version?
Please, don't consider this speach a critical but after the xz archive security incident i'm agree with a lot of distro maintainers that having a way to build from source a critical tool like a package manager that install files in the system it's something important.

Thanks in advance for feedback and thanks to this tool, open source projects slowly became few but you are here and it's really lovely.

**Reply by zkochan** [reply]
You can see here which exact previous version of pnpm is used to build the current version of pnpm: https://github.com/pnpm/pnpm/blob/3bf5e218a6de10a3e56f5a42d6a4a16a0e19a26a/package.json#L66

We were using pnpm to build pnpm since 2016, I believe.

**Reply by zkochan** [reply]
if it helps, we can release pnpm with additional meta information. Like what pnpm version was used to build it.

**Reply by geaaru** [reply]
> You can see here which exact previous version of pnpm is used to build the current version of pnpm:
> 
> https://github.com/pnpm/pnpm/blob/3bf5e218a6de10a3e56f5a42d6a4a16a0e19a26a/package.json#L66
> 
> We were using pnpm to build pnpm since 2016, I believe.

So, how was build the releases before 2016? I know that this means return back to the past but it sounds interesting to know how all is started. For example if it's been used a patched version of npm, etc. I think that could be a good note to add in your website.

**Reply by geaaru** [reply]
> if it helps, we can release pnpm with additional meta information. Like what pnpm version was used to build it.

It sounds like a good idea. :+1:
